### ISSUE TITLE
Feature Request? Working with FOSRestBundle (or any other non-Twig solution)

### ISSUE THREAD

**Issue by lbayerl** [scheb/two-factor-bundle#31]
Is there any way to make this bundle work with FOSRestBundle (or not relying on twig/UI generally)?

**Comment by scheb** [comment]
I haven't worked with FOSRestBundle so far. Can you explain a little bit further what's the problem?

**Comment by Spomky** [comment]
@lbayerl I am not sure this is a good idea.
FOSRestBundle is a bundle that helps to create a resource server
This bundle adds a security layer to authenticate users, not clients. It should be used on an authorization server instead of your API.

**Comment by lbayerl** [comment]
@Spomky - not sure if I got you. We've got an architecture which doesn't seem to be uncommon: A backend providing an REST API (built on Symfony2, Doctrine, FOSUser, FOSRest and JMSSerializer). A frontend built with Angular consuming this UI. Difference to a standard Symfony project might be not to use TWIG and PHP for frontend but decoupling it. So we don't have the client (the full UI) authentificating but every user authentificating (solved by JWT) to the backend and frontend.

User ---Name/Password--> Frontend --Name/Password per JSON to REST Api backend/login --> Backend checks --> Backend provides JWT per JSON ---> Frontend --> User authentificated 

The only thing I am looking for is basically to introduce a second step (second factor) before user is fully authentificated.

So basically I'd like to replace the two-factor-bundle twig template by a JSON request/response which I can expose to my frontend users.

What are your concerns?

**Comment by scheb** [comment]
You could also render JSON with the Twig template. Just saying.

To fully support your case, we'd need to decouple `TwoFactorProvider` from the template engine by introducing some `ResponseRenderInterface` and moving the template rendering logic there. Then, you can replace the renderer with your own implementation by overriding the service configuration (or we even could support exchanging the renderer from config).

**Comment by lbayerl** [comment]
I am convinced it would make sense to decouple the plain functionality from twig, e.g. by the way @scheb proposed.

**Comment by cjgordon** [comment]
@lbayerl this is pretty straight forward, in my FOSRestController:

     $authenticator = $this->container->get('scheb_two_factor.security.google.code_validator');
     if ($authenticator->checkCode($user, $code)) {
            // Do the top secret things here 
     } 


I return AccessDeniedException if it fails. $code is just parameter defined in controller action like any other.
